### PR TITLE
Fast text compat

### PR DESCRIPTION
**Pull request by sebpuetz**
Implement fastText reader and writer.

-------

I'm using `-1` as default for metadata values that we can't infer from the model. The implementation allows serialization of models with FastTextVocab and SimpleVocab.

I'm not 100% sold on providing `->fastText` conversions because it's not lossless and I don't know if fastText is compatible with all the dummy values I threw in. The benefit is opening an avenue for structgram/dependency embeddings in fastText format - which can't be marked as such because the known model types are limited in fastText..

### CODE REVIEW

File: src/finalfusion/compat/fasttext.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        Output file
+    embeds : Embeddings
+        Embeddings to write
+    """
+    with open(file, 'wb') as outf:
+        if not isinstance(embeds.vocab, (FastTextVocab, SimpleVocab)):
+            raise ValueError(
+                f'Expected FastTextVocab or SimpleVocab, not: {type(embeds.vocab).__name__}'
+            )
+        _write_binary(outf, "<ii", 793_712_314, 12)
```
Comment (danieldk) [inline src/finalfusion/compat/fasttext.py]: Use a named constant

File: src/finalfusion/compat/fasttext.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            _write_ft_storage_subwords(outf, embeds)
+        _serialize_array_as_le(outf, embeds.storage)
+
+
+def _read_ft_header(file: BinaryIO):
+    """
+    Helper method to verify version and magic.
+    """
+    magic, version = _read_required_binary(file, "<ii")
+    if magic != 793_712_314:
```
Comment (danieldk) [inline src/finalfusion/compat/fasttext.py]: Named constant

File: src/finalfusion/compat/fasttext.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+def load_fasttext(file: Union[str, bytes, int, PathLike]) -> Embeddings:
+    """
+    Read embeddings from a file in fastText format.
+
+    The returned embeddings have a FastTextVocab, NdArray storage and a Norms chunk.
+
+    Loading embeddings with this method will precompute embeddings for each word by averaging all
+    of its subword embeddings together with the distinct word vector. Additionally, all precomputed
+    vectores are l2-normalized and the corresponding norms are stored in the Norms. The subword
```
Comment (danieldk) [inline src/finalfusion/compat/fasttext.py]: vectores -> vectors

**Review by danieldk — APPROVED** [review]
Some small issues, good otherwise

File: src/finalfusion/compat/fasttext.py (hunk 2 above)
Comment (sebpuetz) [reply src/finalfusion/compat/fasttext.py]: Defining file magic not as a magic number, the irony! Fixed tho ;)

**Comment by sebpuetz** [comment]
Thanks! I'll merge it once the checks are done.

**Comment by danieldk** [comment]
I guess structgram in fastText would only be possible if finalfrontier also supported that format, because of the hashing incompatbilities?

**Comment by sebpuetz** [comment]
> I guess structgram in fastText would only be possible if finalfrontier also supported that format, because of the hashing incompatbilities?

I don't remember if we ever added support for the fastText indexer in finalfrontier. Would be a trivial change to make anyways...

**Comment by sebpuetz** [comment]
https://github.com/finalfusion/finalfrontier/blob/567dc84ac76e24f2a10c4a1cf57d6d6958ec17f4/src/vocab/subword.rs#L221

Looks like we're only missing a line here and changes to arguments

**Comment by danieldk** [comment]
> https://github.com/finalfusion/finalfrontier/blob/567dc84ac76e24f2a10c4a1cf57d6d6958ec17f4/src/vocab/subword.rs#L221
> 
> Looks like we're only missing a line here and changes to arguments

It would be fairly simple indeed!

**Pull request merged**